File: .agents/skills/codebase-intelligence/scripts/python/aggregate.py
#!/usr/bin/env python3
"""Python aggregator — merges the six layer JSONs into raw/python/aggregate.json.

Reads:
    <out-dir>/raw/python/dead_code.json
    <out-dir>/raw/python/duplication.json
    <out-dir>/raw/python/complexity.json
    <out-dir>/raw/python/dependencies.json
    <out-dir>/raw/python/boundaries.json
    <out-dir>/raw/python/hotspots.json

Writes:
    <out-dir>/raw/python/aggregate.json   — Python section for the top-level aggregator

Output shape:
    {
      "lang": "python",
      "python_files": N,
      "python_loc": N,
      "summary": { ...all summary fields... },
      "findings": { ...all findings fields... },
      "warnings": [...]
    }

Note: verdict, schema_version, and repo are NOT included here — those belong
to the top-level aggregate_all.py aggregator.
"""
from __future__ import annotations
import argparse
import json
import subprocess
import sys
from pathlib import Path


def load(path: Path) -> dict:
    if not path.exists():
        return {}
    try:
        return json.loads(path.read_text())
    except json.JSONDecodeError:
        return {"error": True, "reason": "invalid JSON"}


def safe_get(d: dict, *keys, default=None):
    cur = d
    for k in keys:
        if not isinstance(cur, dict) or k not in cur:
            return default
        cur = cur[k]
    return cur


def count_python_loc(repo: Path) -> tuple[int, int]:
    """Count .py files and total LOC (excluding common vendored dirs)."""
    skip = {".venv", "venv", "build", "dist", "node_modules", "__pycache__", ".git"}
    files = 0
    loc = 0
    for p in repo.rglob("*.py"):
        if any(part in skip for part in p.parts):
            continue
        files += 1
        try:
            loc += sum(1 for _ in p.open(encoding="utf-8", errors="ignore"))
        except OSError:
            pass
    return files, loc


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("out_dir")
    ap.add_argument("--repo", required=True)
    ap.add_argument("--scope", default="full")
    args = ap.parse_args()

    out_dir = Path(args.out_dir).resolve()
    repo = Path(args.repo).resolve()
    raw_python = out_dir / "raw" / "python"
    raw_python.mkdir(parents=True, exist_ok=True)

    dead_code = load(raw_python / "dead_code.json")
    duplication = load(raw_python / "duplication.json")
    complexity = load(raw_python / "complexity.json")
    dependencies = load(raw_python / "dependencies.json")
    boundaries = load(raw_python / "boundaries.json")
    hotspots = load(raw_python / "hotspots.json")

    # Collect warnings from skipped/errored layers
    warnings: list[str] = []
    for name, layer in [
        ("dead_code", dead_code), ("duplication", duplication),
        ("complexity", complexity), ("dependencies", dependencies),
        ("boundaries", boundaries), ("hotspots", hotspots),
    ]:
        if layer.get("error"):
            warnings.append(f"layer `{name}` failed — see raw/python/{name}.err")
        if layer.get("skipped"):
            warnings.append(f"layer `{name}` skipped: {layer.get('reason', 'unknown')}")

    py_files, py_loc = count_python_loc(repo)
    duplicated_lines = safe_get(duplication, "counts", "duplicated_lines", default=0)
    dup_rate = round(100.0 * duplicated_lines / py_loc, 2) if py_loc else 0.0

    # Count critical complexity (CC >= 25)
    critical = sum(
        1 for item in complexity.get("findings", [])
        if item.get("verdict") == "above_threshold"
    )

    dep_findings = dependencies.get("findings", [])
    unused_deps = sum(1 for d in dep_findings if d.get("kind") == "unused")
    missing_deps = sum(1 for d in dep_findings if d.get("kind") == "missing_from_pyproject")

    boundary_findings = boundaries.get("findings", [])
    boundary_violations = sum(1 for b in boundary_findings if b.get("kind") == "boundary_violation")
    circular_cycles = sum(1 for b in boundary_findings if b.get("kind") == "circular_import")

    hotspot_top = [
        {"path": h["path"], "score": h["score"]}
        for h in hotspots.get("findings", [])[:5]
    ]

    summary = {
        "dead_code_count": dead_code.get("counts", {}).get("total", 0),
        "duplication_clone_groups": duplication.get("counts", {}).get("clone_groups", 0),
        "duplicated_lines": duplicated_lines,
        "duplication_rate_pct": dup_rate,
        "functions_above_cc_threshold": complexity.get("counts", {}).get("total", 0),
        "functions_at_critical_complexity": critical,
        "avg_maintainability_index": safe_get(complexity, "summary", "avg_maintainability_index"),
        "maintainability_grade": safe_get(complexity, "summary", "maintainability_grade", default="N/A"),
        "unused_dependencies": unused_deps,
        "missing_dependencies": missing_deps,
        "boundary_violations": boundary_violations,
        "circular_import_cycles": circular_cycles,
        "hotspots_top_risk": hotspot_top,
    }

    result = {
        "lang": "python",
        "python_files": py_files,
        "python_loc": py_loc,
        "summary": summary,
        "findings": {
            "dead_code": dead_code.get("findings", []),
            "duplication": duplication.get("findings", []),
            "complexity": complexity.get("findings", []),
            "dependencies": dependencies.get("findings", []),
            "boundaries": boundaries.get("findings", []),
            "hotspots": hotspots.get("findings", []),
        },
        "warnings": warnings,
    }

    (raw_python / "aggregate.json").write_text(json.dumps(result, indent=2))

    print(f"python aggregate: {py_files} files, {py_loc} LOC, "
          f"dead_code={summary['dead_code_count']}, "
          f"critical_cc={critical}, "
          f"missing_deps={missing_deps}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .agents/skills/codebase-intelligence/scripts/typescript/aggregate.js
#!/usr/bin/env node
// aggregate.js — merges all TypeScript layer JSON results into a single summary
// Usage: node aggregate.js <out_dir> --repo <abs-path>

'use strict';

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// --- arg parsing ---
// First positional arg is out_dir; then --repo <path>
const argv = process.argv.slice(2);
let outDir = null;
let repoPath = null;

// Separate positional args from flags
const positional = [];
for (let i = 0; i < argv.length; i++) {
  if (argv[i] === '--repo') {
    repoPath = argv[++i];
  } else if (!argv[i].startsWith('--')) {
    positional.push(argv[i]);
  }
}

outDir = positional[0] || null;

if (!outDir || !repoPath) {
  const msg = { lang: 'typescript', error: true, error_message: 'Usage: node aggregate.js <out_dir> --repo <abs-path>' };
  process.stdout.write(JSON.stringify(msg, null, 2) + '\n');
  process.exit(0);
}

const rawDir = path.join(outDir, 'raw', 'typescript');
fs.mkdirSync(rawDir, { recursive: true });

// --- count TS/JS files and LOC ---
let typescriptFiles = 0;
let typescriptLoc = 0;

try {
  const findCmd = [
    'find',
    repoPath,
    '(',
    '-name', '"*.ts"',
    '-o', '-name', '"*.tsx"',
    '-o', '-name', '"*.js"',
    '-o', '-name', '"*.jsx"',
    ')',
    '-not', '-path', '"*/node_modules/*"',
    '-not', '-path', '"*/dist/*"',
    '-not', '-path', '"*/build/*"',
    '-not', '-path', '"*/.next/*"',
    '-not', '-path', '"*/coverage/*"',
  ].join(' ');

  const findOutput = execSync(findCmd, {
    cwd: repoPath,
    encoding: 'utf8',
    stdio: ['pipe', 'pipe', 'pipe'],
  });

  const files = findOutput.trim().split('\n').filter(Boolean);
  typescriptFiles = files.length;

  // Count lines with wc -l; feed file list via xargs to handle large repos
  if (files.length > 0) {
    // Write file list to temp file for xargs
    const tmpList = path.join(outDir, '.ts_file_list.tmp');
    try {
      fs.writeFileSync(tmpList, files.join('\n') + '\n');
      const wcOutput = execSync(`xargs wc -l < "${tmpList}"`, {
        encoding: 'utf8',
        stdio: ['pipe', 'pipe', 'pipe'],
      });
      // wc -l output: "   42 file.ts\n  100 total\n"
      const totalMatch = wcOutput.match(/(\d+)\s+total/);
      if (totalMatch) {
        typescriptLoc = parseInt(totalMatch[1], 10);
      } else if (files.length === 1) {
        // single file, no "total" line
        const singleMatch = wcOutput.trim().match(/^(\d+)/);
        if (singleMatch) typescriptLoc = parseInt(singleMatch[1], 10);
      }
      fs.unlinkSync(tmpList);
    } catch (e) {
      // fallback: count lines manually for smaller repos
      for (const f of files) {
        try {
          const content = fs.readFileSync(f, 'utf8');
          typescriptLoc += content.split('\n').length;
        } catch (_) {
          // skip unreadable files
        }
      }
    }
  }
} catch (e) {
  // non-fatal — leave counts at 0
}

// --- load layer JSON files ---
const layers = ['dead_code', 'duplication', 'complexity', 'dependencies', 'boundaries', 'hotspots'];
const layerData = {};
const warnings = [];

for (const layer of layers) {
  const layerPath = path.join(rawDir, `${layer}.json`);
  if (!fs.existsSync(layerPath)) {
    warnings.push(`Layer '${layer}' result not found at ${layerPath}`);
    continue;
  }
  try {
    const data = JSON.parse(fs.readFileSync(layerPath, 'utf8'));
    if (data.error) {
      warnings.push(`Layer '${layer}' failed: ${data.error_message}`);
    }
    layerData[layer] = data;
  } catch (e) {
    warnings.push(`Failed to parse layer '${layer}': ${e.message}`);
  }
}

// --- helpers ---
function safeFindings(layer) {
  return (layerData[layer] && !layerData[layer].error && Array.isArray(layerData[layer].findings))
    ? layerData[layer].findings
    : [];
}

function safeCounts(layer) {
  return (layerData[layer] && !layerData[layer].error && layerData[layer].counts)
    ? layerData[layer].counts
    : {};
}

function safeSummary(layer) {
  return (layerData[layer] && !layerData[layer].error && layerData[layer].summary)
    ? layerData[layer].summary
    : {};
}

// --- build summary ---
const deadCounts = safeCounts('dead_code');
const dupCounts = safeCounts('duplication');
const dupSummary = safeSummary('duplication');
const ccCounts = safeCounts('complexity');
const ccSummary = safeSummary('complexity');
const depCounts = safeCounts('dependencies');
const boundsCounts = safeCounts('boundaries');
const hotspotFindings = safeFindings('hotspots');

const topRiskHotspots = hotspotFindings
  .slice(0, 10)
  .map(h => ({ path: h.path, score: h.score }));

// Aggregate hotspot warnings
const hotspotSummary = safeSummary('hotspots');
if (Array.isArray(hotspotSummary.warnings)) {
  for (const w of hotspotSummary.warnings) warnings.push(w);
}

const summary = {
  dead_code_count: (deadCounts.total || 0),
  unused_files: (deadCounts.unused_files || 0),
  unused_exports: (deadCounts.unused_exports || 0),
  duplication_clone_groups: (dupCounts.clone_groups || 0),
  duplicated_lines: (dupCounts.duplicated_lines || 0),
  duplication_rate_pct: dupSummary.duplication_rate_pct || null,
  functions_above_cc_threshold: (ccCounts.above_threshold || 0),
  functions_warn_cc: (ccCounts.warn || 0),
  avg_complexity: ccSummary.avg_complexity || null,
  unused_dependencies: (depCounts.unused || 0),
  unlisted_dependencies: (depCounts.unlisted || 0),
  boundary_violations: (boundsCounts.boundary_violations || 0),
  circular_import_cycles: (boundsCounts.circular_imports || 0),
  hotspots_top_risk: topRiskHotspots,
  warnings,
};

const findings = {
  dead_code: safeFindings('dead_code'),
  duplication: safeFindings('duplication'),
  complexity: safeFindings('complexity'),
  dependencies: safeFindings('dependencies'),
  boundaries: safeFindings('boundaries'),
  hotspots: hotspotFindings,
};

const output = {
  lang: 'typescript',
  typescript_files: typescriptFiles,
  typescript_loc: typescriptLoc,
  summary,
  findings,
};

const aggregatePath = path.join(rawDir, 'aggregate.json');
fs.writeFileSync(aggregatePath, JSON.stringify(output, null, 2));
process.stdout.write(`Wrote aggregate to: ${aggregatePath}\n`);
process.exit(0);
